language/move-lang/src/cfgir/liveness/mod.rs

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+mod state;
+
+use super::{
+    absint::*,
+    ast,
+    ast::*,
+    cfg::{BlockCFG, ReverseBlockCFG, CFG},
+};
+use crate::shared::unique_map::UniqueMap;
+use crate::{errors::*, parser::ast::Var, shared::*};
+use state::*;
+use std::collections::{BTreeMap, BTreeSet, VecDeque};
+
+//**************************************************************************************************
+// Entry and trait bindings
+//**************************************************************************************************
+
+type PerCommandStates = BTreeMap<Label, VecDeque<LivenessState>>;
+type InitialStates = BTreeMap<Label, LivenessState>;
+type FinalInvariants = BTreeMap<Label, FinalInvariant<LivenessState>>;
+struct Liveness {
+    states: PerCommandStates,
+}
+
+impl Liveness {
+    fn new(cfg: &super::cfg::ReverseBlockCFG) -> Self {
+        let states = cfg
+            .blocks()
+            .iter()
+            .map(|(lbl, block)| {
+                let init = block.iter().map(|_| LivenessState::initial()).collect();
+                (*lbl, init)
+            })
+            .collect();
+        Liveness { states }
+    }
+}
+
+impl TransferFunctions for Liveness {
+    type State = LivenessState;
+
+    fn execute(
+        &mut self,
+        state: &mut Self::State,
+        label: Label,
+        idx: usize,
+        cmd: &Command,
+    ) -> Errors {
+        state.next_command();
+        command(state, cmd);
+        // set current [label][command_idx] data with the new liveness data
+        let cur_label_states = self.states.get_mut(&label).unwrap();
+        cur_label_states[idx] = state.clone();
+        Errors::new()
+    }
+}
+
+impl AbstractInterpreter for Liveness {}
+
+pub fn refine(
+    _signature: &FunctionSignature,
+    locals: &UniqueMap<Var, SingleType>,
+    cfg: &mut BlockCFG,
+    infinite_loop_starts: BTreeSet<Label>,
+) {
+    let (final_invariants, per_command_states) = {
+        let mut reverse = ReverseBlockCFG::new(cfg, infinite_loop_starts);
+        analyze(&mut reverse)
+    };
+    let initial_states = build_initial_states(cfg, final_invariants);
+    release_dead_refs(locals, initial_states, per_command_states, cfg.blocks_mut());
+}
+
+//**************************************************************************************************
+// Analysis
+//**************************************************************************************************
+
+fn analyze(reverse: &mut ReverseBlockCFG) -> (FinalInvariants, PerCommandStates) {
+    let initial_state = LivenessState::initial();
+    let mut liveness = Liveness::new(reverse);
+    let final_invariants = liveness.analyze_function(reverse, initial_state).unwrap();
+    (final_invariants, liveness.states)
+}
+
+fn build_initial_states(cfg: &BlockCFG, final_invariants: FinalInvariants) -> InitialStates {
```

Comment:
@sblackshear Let me know what you think of this. It is sort of awkward to do, but after doing the actual analysis, I need more forward information to properly release references. In other words, I need to know the liveness information coming *into* the block as opposed to going *out* of the block

Reply:
I'll add a comment, this is the merged pre states from the forward perspective. 

Conceptually you could almost thing of this as the post state for each block (in the reverse traversal). But that post state would be too weak. So I need to build the "pre" states from the forward direction